### PR TITLE
Github CI tests

### PR DESCRIPTION
**Pull request by elizabethmcd**
This PR adds support for Github CI for linting checks and tests of both the Illumina and Nanopore workflows. Review #31 before this PR.

### CODE REVIEW

**Comment by elizabethmcd** [comment]
I can't confirm if these will work I think until they are merged and then run the tests on future commits?

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      description: |
+        Please drag and drop the relevant files here. Create a `.zip` archive if the extension is not allowed.
+        Your verbose log file `.nextflow.log` is often useful _(this is a hidden file in the directory where you launched the pipeline)_ as well as custom Nextflow configuration files.
+
+  - type: textarea
+    id: system
+    attributes:
+      label: System information
+      description: |
+        * Nextflow version _(eg. 21.10.3)_
```
Comment (taylorreiter) [inline .github/ISSUE_TEMPLATE/bug_report.yml:42]:
```suggestion
        * Nextflow version _(eg. 22.10.1)_
```

File: .github/awstest.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        with:
+          workspace_id: ${{ secrets.TOWER_WORKSPACE_ID }}
+          access_token: ${{ secrets.TOWER_ACCESS_TOKEN }}}
+          compute_env: ${{ secrets.TOWER_COMPUTE_ENV }}
+          workdir: s3://${{ secrets.AWS_S3_BUCKET }}/work/metagenomics/work-${{ github.sha }}
+          parameters: |
+            {
+              "outdir": "s3://${{ secrets.AWS_S3_BUCKET }}/metagenomics/results-${{ github.sha }}"
+            }
+          profiles: test_illumina,test_nanopore,aws_tower
```
Comment (taylorreiter) [inline .github/awstest.yml:25]: Oooh can profiles be combined like this? would this submit two different runs to tower, one for illumina and one for nanopore, or is it just one? it would be really cool if `test_illumina` output a different multiqc than `test_nanopore`, but I'm guessing it doesnt.

File: .github/awstest.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      - name: Launch workflow via tower
+        uses: nf-core/tower-action@v3
+        with:
+          workspace_id: ${{ secrets.TOWER_WORKSPACE_ID }}
+          access_token: ${{ secrets.TOWER_ACCESS_TOKEN }}}
+          compute_env: ${{ secrets.TOWER_COMPUTE_ENV }}
+          workdir: s3://${{ secrets.AWS_S3_BUCKET }}/work/metagenomics/work-${{ github.sha }}
+          parameters: |
+            {
+              "outdir": "s3://${{ secrets.AWS_S3_BUCKET }}/metagenomics/results-${{ github.sha }}"
```
Comment (taylorreiter) [inline .github/awstest.yml:23]: do you want the outdir to have something like `test_` in the name as well? it might help if we start running this workflow a bunch on tower with real data

File: .github/ci.yml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        uses: actions/checkout@v3
+
+      - name: Install Nextflow
+        uses: nf-core/setup-nextflow@v1
+        with:
+          version: "${{ matrix.NXF_VER }}"
+
+      - name: Run Nanopore workflow with test data
+        run: |
+          nextflow run ${GITHUB_WORKSPACE} -profile test_nanopore,docker --outdir ./results_nanopore
```
Comment (taylorreiter) [inline .github/ci.yml:61]: can you use a matrix or something here so that you only have to do all of the test setup once?

it looks like the taxprofiler repo just copies like you have: https://github.com/nf-core/taxprofiler/blob/dev/.github/workflows/ci.yml. As does rnaseq: https://github.com/nf-core/rnaseq/blob/master/.github/workflows/ci.yml. So maybe this is fine the way it is!

File: .github/linting.yml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  push:
+  pull_request:
+  release:
+    types: [published]
+
+jobs:
+  EditorConfig:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (taylorreiter) [inline .github/linting.yml:15]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/linting.yml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+      - uses: actions/setup-node@v2
```
Comment (taylorreiter) [inline .github/linting.yml:17]:
```suggestion
      - uses: actions/setup-node@v3
```

File: .github/linting.yml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+      - name: Install editorconfig-checker
+        run: npm install -g editorconfig-checker
+
+      - name: Run ECLint check
+        run: editorconfig-checker -exclude README.md $(find .* -type f | grep -v '.git\|.py\|.md\|json\|yml\|yaml\|html\|css\|work\|.nextflow\|build\|nf_core.egg-info\|log.txt\|Makefile')
+
+  Prettier:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (taylorreiter) [inline .github/linting.yml:28]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/linting.yml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+      - uses: actions/setup-node@v2
```
Comment (taylorreiter) [inline .github/linting.yml:30]:
```suggestion
      - uses: actions/setup-node@v3
```

File: .github/linting.yml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      - name: Install Prettier
+        run: npm install -g prettier
+
+      - name: Run Prettier --check
+        run: prettier --check ${GITHUB_WORKSPACE}
+
+  PythonBlack:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (taylorreiter) [inline .github/linting.yml:41]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/linting.yml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            ## Python linting (`black`) is failing
+            To keep the code consistent with lots of contributors, we run automated code consistency checks.
+            To fix this CI test, please run:
+            * Install [`black`](https://black.readthedocs.io/en/stable/): `pip install black`
+            * Fix formatting errors in your pipeline: `black .`
+            Once you push these changes the test should pass, and you can hide this comment :+1:
+            We highly recommend setting up Black in your code editor so that this formatting is done automatically on save. Ask about it on Slack for help!
+            Thanks again for your contribution!
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+          allow-repeats: false
```
Comment (taylorreiter) [inline .github/linting.yml:61]: do you want to do nf-core linting? something like this:
```
  nf-core:
    runs-on: ubuntu-latest
    steps:
      - name: Check out pipeline code
        uses: actions/checkout@v3

      - name: Install Nextflow
        uses: nf-core/setup-nextflow@v1

      - uses: actions/setup-python@v4
        with:
          python-version: "3.7"
          architecture: "x64"

      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install nf-core

      - name: Run nf-core lint
        env:
          GITHUB_COMMENTS_URL: ${{ github.event.pull_request.comments_url }}
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          GITHUB_PR_COMMIT: ${{ github.event.pull_request.head.sha }}
        run: nf-core -l lint_log.txt lint --dir ${GITHUB_WORKSPACE} --markdown lint_results.md

      - name: Save PR number
        if: ${{ always() }}
        run: echo ${{ github.event.pull_request.number }} > PR_number.txt

      - name: Upload linting log file artifact
        if: ${{ always() }}
        uses: actions/upload-artifact@v3
        with:
          name: linting-logs
          path: |
            lint_log.txt
            lint_results.md
            PR_number.txt
```

File: .github/awstest.yml (hunk 2 above)
Comment (taylorreiter) [reply .github/awstest.yml:25]: @mertcelebi do you think we could do this for the tower full data sets over at seqqc?

**Review by taylorreiter — COMMENTED** [review]
I left some comments, suggestions, and questions. Commenting for now because you need to bump the nextflow version in the README and the nextflow.config (line 180) to match the nextflow version you're running on CI.

File: .github/awstest.yml (hunk 2 above)
Comment (mertcelebi) [reply .github/awstest.yml:25]: Great question! We should test, I am not sure TBH. ie I am not sure how the combined profiles work.

File: subworkflows/local/nanopore_mapping_depth.nf [hunk 11]
```diff
@@ -0,0 +1,41 @@
+// Minimap2 subworkflow to index an assembly and map reads to the assembly
```
Comment (mertcelebi) [inline subworkflows/local/nanopore_mapping_depth.nf:1]: Why are these changes here?

**Comment by mertcelebi** [comment]
@elizabethmcd can you fix this PR so it only has the GitHub CI changes and not the changes from your other PR?

File: .github/ISSUE_TEMPLATE/bug_report.yml (hunk 1 above)
Comment (elizabethmcd) [reply .github/ISSUE_TEMPLATE/bug_report.yml:42]: ah poop I forgot you had to make these corresponding changes

File: .github/awstest.yml (hunk 2 above)
Comment (elizabethmcd) [reply .github/awstest.yml:25]: I'm not sure either, other than I know I would like to test the separate profiles. And yes they do output different multiqc reports, I've already tested that configuration!

File: .github/linting.yml (hunk 10 above)
Comment (elizabethmcd) [reply .github/linting.yml:61]: we have this commented out in the cookiecutter repo at the moment and I'm not sure if it's because it doesn't work or for some other reason - does it work for you?

Comment (taylorreiter) [reply .github/linting.yml:61]: it works on seqqc, but im not sure if it would work with metagenomics. if you have nf-core tools installed locally, I think you could run `nf-core lint` and see if it your pipeline passes. I have a bunch of ignores set up for specific files that I expect to deviate from the linter, so it did take some config, but its also caught some stuff for me in the past that has been useful

Comment (elizabethmcd) [reply .github/linting.yml:61]: I'm going to leave this out for now but I will make an issue to remind me to come back to this

**Comment by elizabethmcd** [comment]
I have fixed the PR so it doesn't have the files from the previous PR to re-review. I've also changed the `awstest.yml` test to have just the `test_illumina` profile. I've looked at other nf-core workflows that are similarly setup with two separate "subworkflows" and they have a general `test` profile that really only tests one of the subworkflows for this purpose. In the future it would be nice to test both, but for the purposes of getting this up and going I think this is fine.

File: .github/awstest.yml (hunk 2 above)
Comment (elizabethmcd) [reply .github/awstest.yml]: for now I've changed so it just has `test_illumina` although I would like to test both in the future, but starting simple for now

**Review by taylorreiter — APPROVED** [review]
LGTM! I checked it against what I do in seqqc so I think it should work, but always a chance I missed something. I'll be excited to see if it works!
